cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -99,31 +100,40 @@ public void run(JavaSparkExecutionContext sec) throws Exception {
   public static final class ReportSparkHandler extends AbstractSparkHttpServiceHandler {
 
     private static final Logger LOG = LoggerFactory.getLogger(ReportSparkHandler.class);
-    private static final Type REPORT_GENERATION_REQUEST_TYPE = new TypeToken<ReportGenerationRequest>() {
-    }.getType();
+    private static final Type REPORT_GENERATION_REQUEST_TYPE = new TypeToken<ReportGenerationRequest>() { }.getType();
     private static final String DEFAULT_LIMIT = "10000";
-    private SparkSession sparkSession;
+    private static final String READ_LIMIT = "readLimit";
+    private static final String START_FILE = "_START";
+    private static final String FAILURE_FILE = "_FAILURE";
+    private static final AtomicBoolean EXECUTOR_SHUTDOWN = new AtomicBoolean();
 
-    public static final String READ_LIMIT = "readLimit";
-    public static final String THREAD_POOL_SIZE = "poolSize";
-    public static final String START_FILE = "_START";
     public static final String REPORT_DIR = "reports";
     public static final String COUNT_FILE = "COUNT";
     public static final String SUCCESS_FILE = "_SUCCESS";
-    public static final String FAILURE_FILE = "_FAILURE";
+
+    private static ExecutorService reportGenerationExecutor;
 
     private int readLimit;
-    private ExecutorService reportGenerationExecutor;
+    private SparkSession sparkSession;
 
     @Override
     public void initialize(SparkHttpServiceContext context) throws Exception {
       super.initialize(context);
       sparkSession = new SQLContext(getContext().getSparkContext()).sparkSession();
       String readLimitString = context.getRuntimeArguments().get(READ_LIMIT);
       readLimit = readLimitString == null ? 10000 : Integer.valueOf(readLimitString);
-      String threadPoolSizeString = context.getRuntimeArguments().get(THREAD_POOL_SIZE);
-      reportGenerationExecutor =
-        Executors.newFixedThreadPool(threadPoolSizeString == null ? 3 : Integer.valueOf(threadPoolSizeString));
+    }
+
+    @Override
+    public void destroy() {
+      super.destroy();
+      if (EXECUTOR_SHUTDOWN.compareAndSet(false, true)) {
```

Comment:
why do you need this variable ?, you can set the executor to null after shutdown

Reply:
What is that?? Why?? You should just have the executor as a static variable in the Spark program (not the handler) and have the handler use it. When the Spark.run returns (you can block the return with a CountDownLatch and trap the interrupt), just shutdown the executor using finally. Alternatively, if clean shutdown is not needed, you can just use daemon thread and don't worry about the shutdown.

Reply:
There's not much can be done in executor shutdown. I think it will be cleaner to cleanup the states of report generation jobs when the spark program restarts, which can also handle the case of a crash without shutdown. There's a JIRA for this to be addressed in a separate PR.